[BUG] [Remote Store + Segment Replication] Discrepancy between main and 2.x branch of Flush() method causing test failures.

**Describe the bug**
The issue starts with this PR:https://github.com/opensearch-project/OpenSearch/pull/8281, This is a backport PR that successfully passed and merged on main but fails for [testPressureServiceStats](https://build.ci.opensearch.org/job/gradle-check/18675/) on 2.x branch.
The test class [SegmentReplicationRemoteStoreIT](https://github.com/opensearch-project/OpenSearch/blob/2e5195bbf0a1053dfd3647dd22071e1c385eb166/server/src/internalClusterTest/java/org/opensearch/remotestore/SegmentReplicationRemoteStoreIT.java#L31) fails on [Flush()](https://github.com/opensearch-project/OpenSearch/blob/2e5195bbf0a1053dfd3647dd22071e1c385eb166/server/src/internalClusterTest/java/org/opensearch/indices/replication/SegmentReplicationIT.java#L866).


**Reason for test Failure on 2.x:**
Test class [SegmentReplicationRemoteStoreIT](https://github.com/opensearch-project/OpenSearch/blob/2e5195bbf0a1053dfd3647dd22071e1c385eb166/server/src/internalClusterTest/java/org/opensearch/remotestore/SegmentReplicationRemoteStoreIT.java#L31) fails on [testPressureServiceStats()](https://github.com/opensearch-project/OpenSearch/blob/2e5195bbf0a1053dfd3647dd22071e1c385eb166/server/src/internalClusterTest/java/org/opensearch/indices/replication/SegmentReplicationIT.java#L866) when flush() is called. When [Flush()](https://github.com/opensearch-project/OpenSearch/blob/2e5195bbf0a1053dfd3647dd22071e1c385eb166/server/src/main/java/org/opensearch/index/engine/InternalEngine.java#L1923) is called on InternalEngine.java:
Step 1: we ensure translog can flush.
Step 2: Then we call [translogManager.getTranslog().rollGeneration()](https://github.com/opensearch-project/OpenSearch/blob/2e5195bbf0a1053dfd3647dd22071e1c385eb166/server/src/main/java/org/opensearch/index/engine/InternalEngine.java#L1958).
Step 3: Then we refresh.
Step 4: we call [translogManager.trimUnreferencedReaders()](https://github.com/opensearch-project/OpenSearch/blob/2e5195bbf0a1053dfd3647dd22071e1c385eb166/server/src/main/java/org/opensearch/index/engine/InternalEngine.java#L1973), which will further make call to [trimUnreferencedReaders()](https://github.com/opensearch-project/OpenSearch/blob/2e5195bbf0a1053dfd3647dd22071e1c385eb166/server/src/main/java/org/opensearch/index/translog/RemoteFsTranslog.java#L366) in RemoteFsTranslog.java  where we again make a call to
[deleteStaleRemotePrimaryTermsAndMetadataFiles()](https://github.com/opensearch-project/OpenSearch/blob/2e5195bbf0a1053dfd3647dd22071e1c385eb166/server/src/main/java/org/opensearch/index/translog/RemoteFsTranslog.java#L412) w[here](https://github.com/opensearch-project/OpenSearch/blob/2e5195bbf0a1053dfd3647dd22071e1c385eb166/server/src/main/java/org/opensearch/index/translog/RemoteFsTranslog.java#L418) we assert if readers are empty and the assertion fails here
```
java.lang.AssertionError: Expected non-empty readers
	at __randomizedtesting.SeedInfo.seed([3E8142758780CC4C]:0)
	at org.opensearch.index.translog.RemoteFsTranslog.deleteStaleRemotePrimaryTermsAndMetadataFiles(RemoteFsTranslog.java:418)
	at org.opensearch.index.translog.RemoteFsTranslog.trimUnreferencedReaders(RemoteFsTranslog.java:388)
	at org.opensearch.index.translog.InternalTranslogManager.trimUnreferencedReaders(InternalTranslogManager.java:388)
	at org.opensearch.index.engine.InternalEngine.flush(InternalEngine.java:1973)
	at org.opensearch.index.shard.IndexShard.flush(IndexShard.java:1398)
	at org.opensearch.action.admin.indices.flush.TransportShardFlushAction.lambda$shardOperationOnPrimary$1(TransportShardFlushAction.java:110)
	at org.opensearch.action.ActionListener.completeWith(ActionListener.java:342)
	at org.opensearch.action.admin.indices.flush.TransportShardFlushAction.shardOperationOnPrimary(TransportShardFlushAction.java:109)
	at org.opensearch.action.admin.indices.flush.TransportShardFlushAction.shardOperationOnPrimary(TransportShardFlushAction.java:63)
	at org.opensearch.action.support.replication.TransportReplicationAction$PrimaryShardReference.perform(TransportReplicationAction.java:1260)
	at org.opensearch.action.support.replication.ReplicationOperation.execute(ReplicationOperation.java:158)
```
The test fails because we are trying to delete StaleRemotePrimaryTermsAndMetadataFiles when there are no translog readers.


**Reason for test not Failing on main branch:**
When [Flush()](https://github.com/opensearch-project/OpenSearch/blob/2c5a22d167460a0a5d2ad61f16914ca309efe427/server/src/main/java/org/opensearch/index/engine/InternalEngine.java#L1811) method of InternalEngine called:
Step 1: we ensure translog can flush.
Step 2: Then we call [translogManager.rollTranslogGeneration()](https://github.com/opensearch-project/OpenSearch/blob/2c5a22d167460a0a5d2ad61f16914ca309efe427/server/src/main/java/org/opensearch/index/engine/InternalEngine.java#L1846), which internally makes a call to [translog.trimUnreferencedReaders()](https://github.com/opensearch-project/OpenSearch/blob/2c5a22d167460a0a5d2ad61f16914ca309efe427/server/src/main/java/org/opensearch/index/translog/InternalTranslogManager.java#L89C13-L89C48)
Step 3: Then we refresh.
Step 4: Then we call [trimUnreferencedReaders()](https://github.com/opensearch-project/OpenSearch/blob/2c5a22d167460a0a5d2ad61f16914ca309efe427/server/src/main/java/org/opensearch/index/engine/InternalEngine.java#L1861)
The test doesn't fail because trimUnreferencedReaders()  is called twice(and generationsTodelete will be empty) and we never make a call to [deleteStaleRemotePrimaryTermsAndMetadataFiles()](https://github.com/opensearch-project/OpenSearch/blob/2e5195bbf0a1053dfd3647dd22071e1c385eb166/server/src/main/java/org/opensearch/index/translog/RemoteFsTranslog.java#L412) in step 4 as we do in 2.x branch.
The difference in main and 2.x branch is only in step2. Where in 2.x branch we call [translogManager.getTranslog().rollGeneration()](https://github.com/opensearch-project/OpenSearch/blob/2e5195bbf0a1053dfd3647dd22071e1c385eb166/server/src/main/java/org/opensearch/index/engine/InternalEngine.java#L1958) but in main branch we call [translogManager.rollTranslogGeneration()](https://github.com/opensearch-project/OpenSearch/blob/2c5a22d167460a0a5d2ad61f16914ca309efe427/server/src/main/java/org/opensearch/index/engine/InternalEngine.java#L1846)


-> In this PR [#3751](https://github.com/opensearch-project/OpenSearch/pull/3751) of main branch changes made to [flush() method in InternalEngine](https://github.com/opensearch-project/OpenSearch/blob/1d2274045b42e129b8c06748f8185d2431aa5336/server/src/main/java/org/opensearch/index/engine/InternalEngine.java#L1846) of calling translogManager.rollTranslogGeneration()
-> But in 2.x backport of this [PR](https://github.com/opensearch-project/OpenSearch/pull/4095) , we did not backport these changes to [flush() method of InternalEngine](https://github.com/opensearch-project/OpenSearch/blob/c71654444159f643632fbd57588cc00c01b8bb7f/server/src/main/java/org/opensearch/index/engine/InternalEngine.java#L1958). 2.x still has translogManager.getTranslog().rollGeneration()

I am not sure why this change is not backported to 2.x. Might be because author thought it's breaking change or just forgot to include this change in backport.

**To Reproduce**
Steps to reproduce the behavior:
1. Go to '...'
2. Click on '....'
3. Scroll down to '....'
4. See error

**Expected behavior**
A clear and concise description of what you expected to happen.

**Plugins**
Please list all plugins currently enabled.

**Screenshots**
If applicable, add screenshots to help explain your problem.

**Host/Environment (please complete the following information):**
 - OS: [e.g. iOS]
 - Version [e.g. 22]

**Additional context**
Add any other context about the problem here.

--- a/server/src/main/java/org/opensearch/index/engine/InternalEngine.java
+++ b/server/src/main/java/org/opensearch/index/engine/InternalEngine.java
@@ -596,7 +596,7 @@ public class InternalEngine extends Engine {
 
     @Override
     public boolean isTranslogSyncNeeded() {
-        return translogManager.getTranslog().syncNeeded();
+        return translogManager.isTranslogSyncNeeded();
     }
 
     @Override
@@ -611,7 +611,7 @@ public class InternalEngine extends Engine {
 
     @Override
     public TranslogStats getTranslogStats() {
-        return translogManager.getTranslog().stats();
+        return translogManager.getTranslogStats();
     }
 
     @Override
@@ -1892,31 +1892,10 @@ public class InternalEngine extends Engine {
         final long localCheckpointOfLastCommit = Long.parseLong(
             lastCommittedSegmentInfos.userData.get(SequenceNumbers.LOCAL_CHECKPOINT_KEY)
         );
-        final long translogGenerationOfLastCommit = translogManager.getTranslog()
-            .getMinGenerationForSeqNo(localCheckpointOfLastCommit + 1).translogFileGeneration;
-        final long flushThreshold = config().getIndexSettings().getFlushThresholdSize().getBytes();
-        if (translogManager.getTranslog().sizeInBytesByMinGen(translogGenerationOfLastCommit) < flushThreshold) {
-            return false;
-        }
-        /*
-         * We flush to reduce the size of uncommitted translog but strictly speaking the uncommitted size won't always be
-         * below the flush-threshold after a flush. To avoid getting into an endless loop of flushing, we only enable the
-         * periodically flush condition if this condition is disabled after a flush. The condition will change if the new
-         * commit points to the later generation the last commit's(eg. gen-of-last-commit < gen-of-new-commit)[1].
-         *
-         * When the local checkpoint equals to max_seqno, and translog-gen of the last commit equals to translog-gen of
-         * the new commit, we know that the last generation must contain operations because its size is above the flush
-         * threshold and the flush-threshold is guaranteed to be higher than an empty translog by the setting validation.
-         * This guarantees that the new commit will point to the newly rolled generation. In fact, this scenario only
-         * happens when the generation-threshold is close to or above the flush-threshold; otherwise we have rolled
-         * generations as the generation-threshold was reached, then the first condition (eg. [1]) is already satisfied.
-         *
-         * This method is to maintain translog only, thus IndexWriter#hasUncommittedChanges condition is not considered.
-         */
-        final long translogGenerationOfNewCommit = translogManager.getTranslog()
-            .getMinGenerationForSeqNo(localCheckpointTracker.getProcessedCheckpoint() + 1).translogFileGeneration;
-        return translogGenerationOfLastCommit < translogGenerationOfNewCommit
-            || localCheckpointTracker.getProcessedCheckpoint() == localCheckpointTracker.getMaxSeqNo();
+        return translogManager.shouldPeriodicallyFlush(
+            localCheckpointOfLastCommit,
+            config().getIndexSettings().getFlushThresholdSize().getBytes()
+        );
     }
 
     @Override
@@ -1955,7 +1934,7 @@ public class InternalEngine extends Engine {
                     )) {
                     translogManager.ensureCanFlush();
                     try {
-                        translogManager.getTranslog().rollGeneration();
+                        translogManager.rollTranslogGeneration();
                         logger.trace("starting commit for flush; commitTranslog=true");
                         commitIndexWriter(indexWriter, translogManager.getTranslog());
                         logger.trace("finished commit for flush");
@@ -2247,8 +2226,8 @@ public class InternalEngine extends Engine {
             }
             failEngine("already closed by tragic event on the index writer", tragicException);
             engineFailed = true;
-        } else if (translogManager.getTranslog().isOpen() == false && translogManager.getTranslog().getTragicException() != null) {
-            failEngine("already closed by tragic event on the translog", translogManager.getTranslog().getTragicException());
+        } else if (translogManager.getTragicExceptionIfClosed() != null) {
+            failEngine("already closed by tragic event on the translog", translogManager.getTragicExceptionIfClosed());
             engineFailed = true;
         } else if (failedEngine.get() == null && isClosed.get() == false) { // we are closed but the engine is not failed yet?
             // this smells like a bug - we only expect ACE if we are in a fatal case ie. either translog or IW is closed by
@@ -2273,7 +2252,7 @@ public class InternalEngine extends Engine {
             return failOnTragicEvent((AlreadyClosedException) e);
         } else if (e != null
             && ((indexWriter.isOpen() == false && indexWriter.getTragicException() == e)
-                || (translogManager.getTranslog().isOpen() == false && translogManager.getTranslog().getTragicException() == e))) {
+                || (translogManager.getTragicExceptionIfClosed() == e))) {
                     // this spot on - we are handling the tragic event exception here so we have to fail the engine
                     // right away
                     failEngine(source, e);
@@ -2381,7 +2360,7 @@ public class InternalEngine extends Engine {
                     logger.warn("Failed to close ReaderManager", e);
                 }
                 try {
-                    IOUtils.close(translogManager.getTranslog());
+                    IOUtils.close(translogManager);
                 } catch (Exception e) {
                     logger.warn("Failed to close translog", e);
                 }

